ocp_resources/user_defined_network.py

```diff
@@ -0,0 +1,483 @@
+        client: Optional[DynamicClient] = None,
+        role: Optional[str] = None,
+        mtu: Optional[int] = None,
+        subnets: Optional[List] = None,
+        join_subnets: Optional[List] = None,
+        ipam_lifecycle: Optional[str] = None,
+        **kwargs,
+    ):
+        """
+        Create and manage UserDefinedNetwork with layer2 configuration
+
+        Args:
+            name (str): The name of the UserDefinedNetwork.
+            namespace (str): The namespace of the UserDefinedNetwork.
+            client (Optional[DynamicClient]): DynamicClient to use.
+            role (Optional[str]): role describes the network role in the pod.
+            mtu (Optional[int]): mtu is the maximum transmission unit for a network.
+            subnets (Optional[List]): subnets are used for the pod network across the cluster.
+            join_subnets (Optional[List]): join_subnets are used inside the OVN network topology.
+            ipam_lifecycle (Optional[str]): ipam_lifecycle controls IP addresses management lifecycle.
+        """
+        super().__init__(
+            name=name,
+            namespace=namespace,
+            client=client,
+            topology=TopologyType.LAYER2,
+            **kwargs,
+        )
+        self.role = role
+        self.mtu = mtu
+        self.subnets = subnets
+        self.join_subnets = join_subnets
+        self.ipam_lifecycle = ipam_lifecycle
+
+    def to_dict(self) -> None:
+        super().to_dict()
+        if not self.yaml_file:
+            self.res.setdefault("spec", {}).setdefault("layer2", {})
+
+            attributes = {
+                "role": self.role,
+                "mtu": self.mtu,
+                "subnets": self.subnets,
+                "joinSubnets": self.join_subnets,
+                "ipamLifecycle": self.ipam_lifecycle,
+            }
+
+            for key, value in attributes.items():
+                if value is not None:
+                    self.res["spec"]["layer2"][key] = value
+
+
+class Layer3Subnets:
+    """
+    UserDefinedNetwork layer3 subnets object.
+
+    API reference:
+    https://ovn-kubernetes.io/api-reference/userdefinednetwork-api-spec/#layer3subnet
+    """
+
+    def __init__(
+        self,
+        cidr: Optional[str] = None,
+        host_subnet: Optional[int] = None,
+    ):
+        """
+        UserDefinedNetwork layer3 subnets object.
+
+        Args:
+            cidr (Optional[str]): CIDR specifies L3Subnet, which is split into smaller subnets for every node.
+            host_subnet (Optional[int]): host_subnet specifies the subnet size for every node.
+        """
+        self.cidr = cidr
+        self.host_subnet = host_subnet
+
+
+class Layer3UserDefinedNetwork(UserDefinedNetwork):
+    """
+    UserDefinedNetwork layer3 object.
+
+    API reference:
+    https://ovn-kubernetes.io/api-reference/userdefinednetwork-api-spec/#layer3config
+    """
+
+    def __init__(
+        self,
+        name: str,
+        namespace: str,
```

Comment:
Ditto (is `namespace` mandatory)?

Reply:
Yes it is mandatory, see my comment:
https://github.com/RedHatQE/openshift-python-wrapper/pull/2124#discussion_r1782679328